File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/list/SwipeDiaryDayListBaseAdapter.java
package com.websarva.wings.android.zuboradiary.ui.list;

import android.content.Context;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.websarva.wings.android.zuboradiary.data.preferences.ThemeColor;

import java.time.LocalDate;
import java.util.Objects;

public abstract class SwipeDiaryDayListBaseAdapter extends DiaryDayListBaseAdapter {

    private OnClickDeleteButtonListener onClickDeleteButtonListener;

    protected SwipeDiaryDayListBaseAdapter(
            Context context,
            RecyclerView recyclerView,
            ThemeColor themeColor,
            DiffUtilItemCallback diffUtilItemCallback) {
        super(context, recyclerView, themeColor, diffUtilItemCallback);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) {
        super.onBindViewHolder(holder, position);

        DiaryDayListBaseItem item = getItem(position);
        Objects.requireNonNull(item);

        onBindDeleteButtonClickListener(holder, item);
    }

    protected abstract void onBindDeleteButtonClickListener(@NonNull RecyclerView.ViewHolder holder, @NonNull DiaryDayListBaseItem item);

    @FunctionalInterface
    public interface OnClickDeleteButtonListener {
        void onClick(LocalDate date);
    }

    public void setOnClickDeleteButtonListener(OnClickDeleteButtonListener onClickDeleteButtonListener) {
        Objects.requireNonNull(onClickDeleteButtonListener);

        this.onClickDeleteButtonListener = onClickDeleteButtonListener;
    }

    protected void onClickDeleteButton(LocalDate date) {
        Objects.requireNonNull(date);
        if (onClickDeleteButtonListener == null) return;

        onClickDeleteButtonListener.onClick(date);
    }
}

File: app/src/main/java/com/websarva/wings/android/zuboradiary/ui/list/SwipeDiaryDayListBaseAdapter.kt
package com.websarva.wings.android.zuboradiary.ui.list

import android.content.Context
import androidx.recyclerview.widget.RecyclerView
import com.websarva.wings.android.zuboradiary.data.preferences.ThemeColor

abstract class SwipeDiaryDayListBaseAdapter protected constructor(
    context: Context,
    recyclerView: RecyclerView,
    themeColor: ThemeColor,
    diffUtilItemCallback: DiffUtilItemCallback
) : DiaryDayListBaseAdapter(context, recyclerView, themeColor, diffUtilItemCallback) {

    fun interface OnClickDeleteButtonListener {
        fun onClick(item: DiaryDayListBaseItem)
    }
    var onClickDeleteButtonListener: OnClickDeleteButtonListener? = null

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        super.onBindViewHolder(holder, position)

        val item = getItem(position)
        onBindDeleteButtonClickListener(holder, item)
    }

    protected abstract fun onBindDeleteButtonClickListener(
        holder: RecyclerView.ViewHolder,
        item: DiaryDayListBaseItem
    )

    protected fun onClickDeleteButton(item: DiaryDayListBaseItem) {
        onClickDeleteButtonListener?.onClick(item) ?: return
    }
}
